src/main/java/sx/blah/discord/util/MessageHistoryBuilder.java

```diff
@@ -0,0 +1,429 @@
+	 *
+	 * @param endTime the time to end at
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(LocalDateTime endTime) {
+		this.uEnd = null;
+		this.uEndTime = endTime;
+		this.uIncludeEnd = false;
+		return this;
+	}
+
+	/**
+	 * Specifies the message to end at.
+	 *
+	 * @param endID     the message ID to end at
+	 * @param inclusive whether or not to include this time
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(long endID, boolean inclusive) {
+		this.uEnd = channel.getMessageByID(endID);
+		if (uEnd == null)
+			throw new IllegalArgumentException("Message " + Long.toUnsignedString(endID) + " does not exist in channel " + channel.getStringID() + "!");
+		this.uEndTime = uEnd.getTimestamp();
+		this.uIncludeEnd = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the message to end at.
+	 *
+	 * @param msg       the message ID to end at
+	 * @param inclusive whether or not to include this time
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(IMessage msg, boolean inclusive) {
+		if (!channel.equals(msg.getChannel()))
+			throw new IllegalArgumentException("Message " + msg.getStringID() + " does not exist in channel " + channel.getStringID() + "!");
+		this.uEnd = msg;
+		this.uEndTime = msg.getTimestamp();
+		this.uIncludeEnd = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the time to end at.
+	 *
+	 * @param endTime   the time to end at
+	 * @param inclusive whether or not to include this time
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder endAt(LocalDateTime endTime, boolean inclusive) {
+		this.uEnd = null;
+		this.uEndTime = endTime;
+		this.uIncludeEnd = inclusive;
+		return this;
+	}
+
+	/**
+	 * Specifies the maximum number of messages to get. Defaults to -1 (no limit).
+	 *
+	 * @param count max number of messages to get
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder withMaxCount(int count) {
+		if (count < 1) throw new IllegalArgumentException();
+		this.count = count;
+		return this;
+	}
+
+	/**
+	 * Returns a stream of the message history.
+	 *
+	 * @return the stream of the message history
+	 * @see MessageHistoryIterator
+	 */
+	public Stream<IMessage> stream() {
+
+		buildRange();
+
+		return new MessageHistoryIterator(range).stream();
+	}
+
+	/**
+	 * Returns an iterator for this message history.
+	 * NOTE: This object can be put directly in a for-each loop.
+	 *
+	 * @return an Iterator instance for the message history
+	 */
+	@Override
+	public Iterator<IMessage> iterator() {
+
+		buildRange();
+
+		return new MessageHistoryIterator(range);
+	}
+
+	/**
+	 * Requests a {@link MessageHistory} object with the parameters set in this builder.
+	 * NOTE: This will not block the thread during execution. Use {@link MessageHistoryBuilder#get()} to block the
+	 * thread.
+	 *
+	 * @return the message history
+	 * @see MessageHistory
+	 */
+	public RequestBuffer.RequestFuture<MessageHistory> request() {
```

Comment:
Seems like the logic for `request()` and `get()` are backwards. `get()` should just be a blocking `request()`.